### ISSUE TITLE
面试准备-杂七杂八

### ISSUE THREAD

**Issue by Eamymao** [Eamymao/winter-ann-arbor#7]
### 神经网络相关
- [x] 最新的词向量领域的模型BERT, ELMO

BERT 和 word2vec的区别

先说他们的区别，最大的区别是，BERT中每个位置的词向量，都要经过一个多层的transformer的网络结构，transformer的网络结构会将每个位置的词向量和其他位置的词向量进行名叫self-attention的矩阵变化，**最终每个位置的词向量将会融合了各个位置的词向量的信息**。而Word2Vec我们得到的只是网络的一个参数矩阵，每个词的表示不会因其所在句子不同而发生改变。因此BERT比Word2Vec进步的地方在于，BERT每个位置经过多次transformer输出后的词向量，是有语境信息，而且比之前的基于RNN的模型BERT能更直接的建模距离更远的词和词之间的依赖关系，这更是Word2Vec所不具备的。

从训练方法上来说，BERT采用Denoise的模式去预测随机MASK掉的位置上的词，由于采用字模型，字的数量相比于词要少，BERT去预测MASK掉的词的时候是直接通过softmax然后使用交叉熵作为损失函数去训练模型。而Word2Vec则一般采用一个滑动窗口去预测窗口中间词或者从中间词预测窗口两边的词，同时由于采用词作为基本单元，词的数量相对较大，Word2Vec一般采用分层softmax或者negative sample的方式去训练。当然，不考虑计算量的问题，直接采用softmax效果是否比negative sample要好，这个不好说了。从训练方法上，两者都有值得借鉴的地方。

从使用者的角度，Word2Vec也是一个预训练的模型，模型的参数就是词向量本身，BERT作为一个更加强大的预训练模型，可以将句子中的词通过Transformer输出为有丰富语境信息的词向量，同时还可以得到句子向量，并且BERT的预训练更加充分。BERT和Word2Vec都可以作为其他模型的一部分，而BERT更强。但是强的代价是BERT的计算效率很低，而Word2Vec就是一个查表的过程，效率非常高，这也限制了BERT在许多要求实时的领域的应用。

- [ ] Transformer的原理，和RNN的对比
https://zhuanlan.zhihu.com/p/48508221
RNN相关算法只能从左向右依次计算或者从右向左依次计算，这种机制带来了两个问题：

> https://zhuanlan.zhihu.com/p/106106079

1. 时间片 [公式] 的计算依赖 [公式] 时刻的计算结果，这样限制了模型的并行能力；
2. 顺序计算的过程中信息会丢失，尽管LSTM等门机制的结构一定程度上缓解了长期依赖的问题，但是对于特别长期的依赖现象,LSTM依旧无能为力。

Transformer的提出解决了上面两个问题，首先它使用了Attention机制，将序列中的任意两个位置之间的距离是缩小为一个常量；其次它不是类似RNN的顺序结构，因此具有更好的并行性，符合现有的GPU框架。

http://jalammar.github.io/illustrated-transformer/ 英文版解释

- [x] LSTM的公式（三个门）；LSTM怎么解决的RNN梯度爆炸/消失问题（如何解决长距离依赖问题）

三个门的工作流程
> https://zhuanlan.zhihu.com/p/38057400

为什么相比于RNN，LSTM在梯度消失上表现更好
> https://www.zhihu.com/question/44895610

> https://www.cnblogs.com/bonelee/p/10475453.html
LSTM 中为什么不使用RELU函数而使用tanh？

> https://www.zhihu.com/question/61265076

- [x] Word2Vec算法复习

- [x] Dropout的原理 (为什么训练时有dropout测试时没有dropout，这样会发生scale的偏移吗)

答：
- Dropout 是在训练过程中以一定的概率的使神经元失活，即输出为0，以提高模型的泛化能力，减少过拟合。
- Dropout 在训练时采用，是为了减少神经元对部分上层神经元的依赖，类似将多个不同网络结构的模型集成起来，减少过拟合的风险。
而在测试时，应该用整个训练好的模型，因此不需要dropout。

- 假设失活概率为 p ，就是这一层中的每个神经元都有p的概率失活，如下图的三层网络结构（6个节点）中，如果失活概率为0.5，则平均每一次训练有3个神经元失活，所以输出层每个神经元只有3个输入，而实际测试时是不会有dropout的，输出层每个神经元都有6个输入，这样在训练和测试时，输出层每个神经元的输入和的期望会有量级上的差异。
因此在训练时还要对第二层的输出数据除以（1-p）之后再传给输出层神经元，作为神经元失活的补偿，以使得在训练时和测试时每一层输入有大致相同的期望。
不会，因为训练时dropout机制会把dropout_rate = p的输出乘以1/(1-p)（这样就可以使结果的scale保持不变，而在预测的时候也不用做额外的操作了，更方便一些）

- [x] Batch Normalization的原理

答：在mini_batch上进行而不是在整个数据集上进行。在训练集上有BN测试集上没有BN，这样不会发生数值上的偏移，道理类似dropout，即记录下训练集上的BN参数（均值、标准差、\beta、\gamma）作为网络参数的一部分。

> https://zhuanlan.zhihu.com/p/34879333

> https://blog.csdn.net/hjimce/article/details/50866313

- [x] 为什么要用ReLU替代tanh/sigmoid/softmax

答：ReLU求导的值要么是0要么是1，不会引起梯度消失或梯度爆炸

- [ ] ResNet

答：通过x + f(x)的结果，让深层网络获得不差于浅层网络的学习能力，从而允许网络变得更深

### 机器学习相关

- [ ] 信息增益，熵和互信息

> https://zhuanlan.zhihu.com/p/35423404

- [x] 逻辑回归的目标函数，损失函数，梯度的推导（最好还会现场编出来）
https://blog.csdn.net/ligang_csdn/article/details/53838743

logit函数与sigmoid函数的联系：互为反函数
logit:
$$logit(p) = \dfrac{p}{1-p} = w^Tx$$
sigmoid:
$$sigmoid(x) = \dfrac{1}{1+e^{-w^Tx}}  $$
- [ ] k-means的算法实现

- [x] 常用优化算法有哪些，分别有什么优缺点

> https://blog.csdn.net/zhouhong0284/article/details/80232412

> https://www.cnblogs.com/shixiangwan/p/7532830.html

- [x] 交叉熵是什么，默写公式

> https://blog.csdn.net/rtygbwwwerr/article/details/50778098
都反映了分布p，q的相似程度
$$-\sum p(x)logq(x)$$

- [x] 损失函数有哪些，损失函数的作用
损失函数(loss function)，或者叫代价函数(cost function)。损失函数越小，就代表模型拟合的越好
目标函数 = 损失函数+结构风险，结构风险函数用来衡量模型的复杂度（正则项）
优化函数是：$$\frac{1}{N}\sum_{i=1}^nL(y_i,f(x_i))+\alpha J(x)$$，即最优化经验风险和结构风险，而这个函数就被称为目标函数

> 常用损失函数：https://blog.csdn.net/weixin_37933986/article/details/68488339

- [ ] self-attention 到底比seq2seq的attention优越在哪里(结合业务场景)
> https://zhuanlan.zhihu.com/p/46250529
> https://zhuanlan.zhihu.com/p/47470866

- [x] L1，L2正则化区别，有什么作用，公式简述

> https://www.zhihu.com/question/26485586

- [ ] 问了个抖音业务相关的问题，follow行为的预测

- [x] GBDT和xgboost和随机森林的区别

> https://blog.csdn.net/a819825294/article/details/51206410 XGB原理

> https://www.zhihu.com/question/60625492 梯度提升树中为什么说目标函数关于当前模型的负梯度是残差的近似

> https://www.cnblogs.com/pinard/p/6140514.html GBDT原理

> https://zhuanlan.zhihu.com/p/88946433 由adboost, gbdt到xgboost，从目标函数说起

> https://www.zhihu.com/question/41354392

> https://zhuanlan.zhihu.com/p/29765582

- [ ] boosting和bagging的区别

- [ ] 机器学习模型优化中常见问题和解决思路
https://blog.csdn.net/mozhizun/article/details/71438821
